### PR TITLE
Handling token expiration

### PR DESCRIPTION
**Pull request by vishalnarkhede**
# Submit a pull request

### Static token

```js
client.setUser({id: 'vishal'}, 'token');
```

### Token provider

```js
client.setUser({id: 'vishal'}, async () => {
    // const token = await api.fetchToken();
    return token;
});
```

## Description

- New class - `TokenManager`, which will be a common place to access token from for client and connection class.
- When ws connection fails with error code `40`, the token will be refreshed using tokenProvider function (given to `setUser` function) and connection request will be retried.
- When http request fails with error code `40`, the token will be refreshed using tokenProvider function (given to `setUser` function) and connection request will be retried.

## TODO

- [x] Handle error from ws connection about expired token
- [x] Handle error from http request about expired token
- [x] Handle the case of guest users
- [x] Handle the case of anonymous users
- [x] tests
- [ ] docs

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -472,93 +468,65 @@ export class StreamChat {
 		});
 	}
 
-	async get(url, params) {
+	doAxiosRequest = async (type, url, data, params = {}) => {
```
Comment (vishalnarkhede) [inline src/client.js:471]: Keeping all the logging and response handling in a single place!

File: src/client.js [hunk 2]
```diff
@@ -195,30 +194,16 @@ export class StreamChat {
...
-		}
+		await this._setToken(user, userTokenOrProvider);
 		this._setUser(user);
 		this.anonymous = false;
 
 		return this._setupConnection();
-	}
+	};
+
+	_setToken = async (user, userTokenOrProvider) => {
```
Comment (tbarbugli) [inline src/client.js:204]: ideally we add some validation for `userTokenOrProvider` just to make sure that you dont pass silly stuff to it

File: src/client.js [hunk 3]
```diff
@@ -323,6 +308,8 @@ export class StreamChat {
 		this.activeChannels = {};
 		// reset client state
 		this.state = new ClientState();
+		// reset token manager
+		this.tokenManager.expire();
```
Comment (tbarbugli) [inline src/client.js:312]: why dont we call this `reset` then?

File: src/client.js (hunk 2 above)
Comment (tbarbugli) [reply src/client.js:204]: isn't pointless to have this as an async fn just to await on `this.tokenManager.setTokenOrProvider`

File: src/client.js [hunk 4]
```diff
@@ -472,93 +465,68 @@ export class StreamChat {
...
-
-			return this.handleResponse(response);
-		} catch (e) {
-			this._logApiError('post', url, e);
 			if (e.response) {
+				if (
+					e.response.data.code === chatCodes.TOKEN_EXPIRED &&
+					!this.tokenManager.isStatic()
+				) {
+					await this.tokenManager.loadToken();
```
Comment (tbarbugli) [inline src/client.js:502]: one simple optimization available here:

```js

function doAxiosRequest() {
   await this.tokenManager.tokenReady(); // return a promise from loadToken (resolved if finished, pending if running)
   // the actual request handling

   if (e.response.data.code === chatCodes.TOKEN_EXPIRED && !this.tokenManager.isStatic()) {
      return await this.doAxiosRequest(type, url, params, data);
   }
}
```

File: src/client.js (hunk 2 above)
Comment (vishalnarkhede) [reply src/client.js]: > ideally we add some validation for `userTokenOrProvider` just to make sure that you dont pass silly stuff to it

We have validation in token_manager actually

**Comment by ferhatelmas** [comment]
@vishalnarkhede when I say this way, you might not like much but there are 3 type of changes here:
* syntax due to new proposals
* refactoring for common error handling
* token expiration feature

It would be better if you could extract first two items into its own PR then we're left alone with the feature so that the details can reviewed better.

**Comment by vishalnarkhede** [comment]
@ferhatelmas 

> syntax due to new proposals

??

**Comment by ferhatelmas** [comment]
for example, https://github.com/GetStream/stream-chat-js/pull/319/files#diff-cf27c1d543e886c89cd9ac8b8aeaf05bR158 and also with some formatting changes.

**Comment by vishalnarkhede** [comment]
@ferhatelmas I hear you and I will be careful in next PR. But it's not always possible to isolate every little change. All of them were done to implement token expiration functionality (except for the one that you mentioned in last comment). At this point it would be too much effort to extract them to separate PR!

**Comment by ferhatelmas** [comment]
Thanks. If we slowly adapt keeping changes to be independent, it would be useful when problems are seen and some investigation is needed.

OK, don't worry about this.

**Pull request merged**